### ISSUE TITLE
ActionCable doesn't reconnect / update after cookies are changed [Chrome / Firefox only]

### ISSUE THREAD

**Issue by ideo-henry** [rails/rails#25882]
### Steps to reproduce
In production mode, when a user is logged in and changes the cookies, the actioncable :identified_by doesn't get updated.

Here is how the cookies are updated (student users are logged in):
```ruby
	def create
		session[:student_id] = nil
		cookies.delete :student_id
		@student = Student.find_by_email(params[:student_session][:email])
		if @student && @student.authenticate(params[:student_session][:password])
			session[:student_id] = @student.id
			cookies.signed[:student_id] = @student.id
			redirect_to '/student_space'
		else
			redirect_to '/student_login'
		end
	end
```

Here is the actioncable code:
```ruby
module ApplicationCable
  class Connection < ActionCable::Connection::Base
  	identified_by :current_student, :current_tutor

  	def connect
      if find_verified_user != nil
        self.current_student = find_verified_user[0]
  		  self.current_tutor = find_verified_user[1]
      end
  	end

  	private
  		def find_verified_user
  			if (current_student = Student.find_by(id: cookies.signed[:student_id])) || (current_tutor = Tutor.find_by(id: cookies.signed[:tutor_id]))
  				return [current_student, current_tutor]
  			#else
  				#reject_unauthorized_connection
  			end
  		end
  end
end
```

The subscribing part:
```ruby
  def subscribed
  	if (current_student != nil)
    	stream_from "chat_channel_#{current_student.id}"
    elsif (current_tutor != nil)
    	current_tutor.chats.each do |chat|
    		stream_from "chat_channel_#{chat.channel_id}"
    	end
   	end
  end
```

### Expected behavior
The actioncable should re-connect and update where it's streaming from when cookies are changed.

### Actual behavior
In production mode, and in Chrome / Firefox only, the connection doesn't get updated when the cookies are changed, thus, it's not streaming from the right streams.

### System configuration
**Rails version**: 5.0.0

**Ruby version**: 2.3

**Comment by maclover7** [comment]
Hmm, can you condense down the code above into a more concise reproduction script? Also, is the Action Cable connection already established before the cookies are changed? In that case, it's already too late for the new information to be used.

**Comment by ideo-henry** [comment]
Thank you, @maclover7! The code is now minimized. I'm not sure whether the Action Cable connection has already established. When the cookie is changed, I ```redirect_to``` a different page. Therefore, I guess it should establish a new connection? If that's the case, shouldn't Rails establish a new connection based on the new cookie?

**Comment by danielrhodes** [comment]
@ideo-henry No, I don't think that would be a desirable behavior, even in your case.

Right now there is definitely no reconnect on a cookie change. Since this is an open socket, for as long as the connection is established any change in the client state (cookies, etc.) is not going to be transmitted - the only state the server knows about is when the connection was initially made. This is the same behavior you would see with long-polling or SSE or any other method of keeping a connection alive (HTTP2 might be different, would have to investigate).

If you disconnect and connect again, however, the browser should be sending the new cookies just as if you would make  an Ajax request after changing the cookies. This is what I would recommend.

In my opinion, this re-connect is not something that should be done automatically - it should be done manually by the developer. Sockets encompass a different style of interaction with a server, and you definitely don't want your connection to be flapping or reconnecting for reasons that are not an explicit side effect of something that seems like a benign change.

**Comment by ideo-henry** [comment]
@danielrhodes Thank you so much for pointing this out! How can I explicitly re-connect to rails actioncable from front-end JavaScript calls and / or back-end rails code?

**Comment by danielrhodes** [comment]
Not sure if it's in the guides, but here is some documentation:

https://github.com/rails/rails/blob/master/actioncable/app/assets/javascripts/action_cable/consumer.coffee

**Comment by ideo-henry** [comment]
@danielrhodes Thank you for the documentation! However, after I've added this JavaScript code here to re-connect to the actioncable

```javascript
	$(document).ready(function(){
        setTimeout(function(){
            App.cable.disconnect();
        }, 250);
        setTimeout(function(){
            App.cable.connect();
        }, 500);
    });
```

The issue still remains. In Chrome and Firefox, the action-cable is still using the old cookies after re-connecting. It seems that the server can receive the data from action-cable but the client end is not streaming from the correct streams due to the outdated cookies, and ```received``` is never called.

Could this be a browser-specific issue? This works fine on Safari and IE.

**Comment by danielrhodes** [comment]
Eek, I don't have a quick answer for you there. I would verify that when you reconnect, it is indeed sending the old cookie. In the Developer console, you can see the request which turns into a web socket (should be to something like `/cable`) and also you can see the websocket frames.

**Comment by merqlove** [comment]
Yep, it's because Consumer object still alive.

I'm doing something like cleaning all subscriptions, then closing socket, dropping consumer object & then disconnect/connect with new credentials.
